bin/action_runner.py

```diff
@@ -5,79 +5,96 @@
 
-opener = open
-
-
 class StatusFile(object):
     """Manage a status file."""
 
     def __init__(self, filename):
         self.filename = filename
 
-    def write(self, command, proc):
-        with opener(self.filename, 'w') as fh:
-            yaml.dump(self.get_content(command, proc), fh)
-
-    def get_content(self, command, proc):
+    def get_content(self, run_id, command, proc):
         return {
+            'run_id':       run_id,
             'command':      command,
             'pid':          proc.pid,
             'return_code':  proc.returncode,
         }
 
     @contextlib.contextmanager
-    def wrap(self, command, proc):
-        self.write(command, proc)
-        try:
-            yield
-        finally:
-            self.write(command, proc)
-
-
-class NoFile(object):
-
-    @classmethod
-    @contextlib.contextmanager
-    def wrap(self, _command, _proc):
-        yield
+    def wrap(self, command, run_id, proc):
+        with open(self.filename, 'w') as fh:
+            yaml.safe_dump(
+                self.get_content(
+                    run_id=run_id,
+                    command=command,
+                    proc=proc,
+                ), fh,
+            )
+            try:
+                yield
+            finally:
+                yaml.safe_dump(
+                    self.get_content(
+                        run_id=run_id,
+                        command=command,
+                        proc=proc,
+                    ), fh,
+                )
 
 
 def get_status_file(output_path):
-    if not os.path.isdir(output_path):
+    if os.path.isdir(output_path):
+        if not os.access(output_path, os.W_OK):
+            raise OSError("Output dir %s not writable" % output_path)
+        return StatusFile(os.path.join(output_path, STATUS_FILE))
+    else:
         try:
             os.makedirs(output_path)
         except OSError:
-            log.warn("Output path %s does not exist", output_path)
-            return NoFile
-    return StatusFile(os.path.join(output_path, STATUS_FILE))
+            raise OSError("Could not create output dir %s" % output_path)
+        return StatusFile(os.path.join(output_path, STATUS_FILE))
 
 
-def register(output_path, command, proc):
+def run_proc(output_path, command, run_id, proc):
     status_file = get_status_file(output_path)
-    with status_file.wrap(command, proc):
+    with status_file.wrap(
+        command=command,
+        run_id=run_id,
+        proc=proc,
+    ):
         proc.wait()
     sys.exit(proc.returncode)
 
 
-def parse_args(args):
-    if len(args) != 3:
-        raise SystemExit("Requires both output_path and command.")
-    return args[1:]
+def parse_args():
+    parser = argparse.ArgumentParser(description='Action Runner for Tron')
+    parser.add_argument(
+        'output_dir',
+        help='an integer for the accumulator',
+    )
+    parser.add_argument(
+        'command',
+        help='the command to run',
+    )
+    parser.add_argument(
```

Comment:
Is this immune to "quoting" issues? Before the command was the "last" argument, it it OK to have the ID as the last argument now, and is it safe to assume the "command" is properly contained as the "second argument".

The commands passed to tron are very complex and have very complex quoting. Here of all places we need to be super-careful that the optparse-argparse transition is safe.

Reply:
I don't think its the command parser that causes the problem, just that we don't quote correctly in actioncommand.py where we generate the args to action_runner (see TRON-213 for an example of it being broken right now).

I've fixed this by quoting everything in actioncommand.py and tested that the broken example in TRON-213 is fixed (as well as adding a test that splits the command generated, as argparse would).